### PR TITLE
manually hash `TyKind`

### PR DESCRIPTION
**Pull request by llogiq**
This might improve perf somewhat. Fingers crossed.

### CODE REVIEW

File: compiler/rustc_middle/src/ty/sty.rs [hunk 1]
```diff
@@ -198,6 +198,107 @@ pub enum TyKind<'tcx> {
...
 
+// try to use less hash operations to speed up hashing compared to a derived
+// implementation
+impl<'tcx> std::hash::Hash for TyKind<'tcx> {
+    fn hash<H: std::hash::Hasher>(&self, h: &mut H) {
+        let disc = std::intrinsics::discriminant_value(self) as usize;
+        match self {
+            Self::Bool | Self::Char | Self::Str | Self::Never | Self::Error(_) => disc.hash(h),
+            Self::Int(i) => {
+                (disc | ((std::intrinsics::discriminant_value(i) as usize) << 5)).hash(h);
```
Comment (the8472) [inline compiler/rustc_middle/src/ty/sty.rs:209]: Maybe the 5 should be extracted to a `const` with a comment that it needs to be updated if the number of variants changes?

Comment (llogiq) [reply compiler/rustc_middle/src/ty/sty.rs:209]: Agreed. I wanted a perf run before I take care of such details though.

Comment (camelid) [reply compiler/rustc_middle/src/ty/sty.rs:209]: Even better, you could just use [`std::mem::variant_count`](https://doc.rust-lang.org/nightly/std/mem/fn.variant_count.html).

**Comment by llogiq** [comment]
Results seem to be a wash, not really worth it. Probably the shift trick in conjunction with `DefId`s and similar types reduces hash quality. Perhaps going a more conservative route and only combining the integer types will be better.

**Comment by llogiq** [comment]
While I was at it, I also implemented the variant_count based `const` shift.

Can I have another perf run?

**Comment by matthiaskrgr** [comment]
sure :)
@bors try @rust-timer queue

**Comment by camsteffen** [comment]
@llogiq I believe you're able to do perf runs because I granted it to the clippy team a while ago :)

**Comment by llogiq** [comment]
Going to try that next time.

**Comment by llogiq** [comment]
Meh, that doesn't appear to be worth the added complexity. Closing this.

**Pull request closed without merging**